File: src/strings/easy/first-non-repeating-char/go/map.go
package main

// O(n) time | O(1) space
// where N is the length of the input string
// the space stays constant cause we will at most store 26 keys (or whatever our constant alphabet is)
func FirstNonRepeatingCharacter(str string) int {
	charFrequency := map[rune]int{}

	for _, char := range str {
		charFrequency[char] += 1
	}

	for i, char := range str {
		if charFrequency[char] == 1 {
			return i
		}
	}

	return -1
}

File: src/strings/easy/first-non-repeating-char/js/map.js
// O(n) time | O(1) space - where N is the length of the input string
// The input string will only have lowercase letters so 26 at most
function firstNonRepeatingCharacter(string) {
    const charFrequencies = {}

    for (const char of string) {
        if (!(char in charFrequencies)) {
            charFrequencies[char] = 0
        }
        charFrequencies[char]++
    }

    for (let i = 0; i < string.length; i++) {
        const char = string[i]
        if (charFrequencies[char] === 1) {
            return i
        }
    }

    return -1
}

exports.firstNonRepeatingCharacter = firstNonRepeatingCharacter;
